### ISSUE TITLE
Trait searches can return traits that don't have a GWAS or a QTL study

### ISSUE THREAD

**Issue by alancleary** [legumeinfo/graphql-server#104]
If I recall correctly, playing with the MiniMine QueryBuilder while working on the GraphQL `traits` query revealed that you can get results that don't have a value for `gwas` or `qtlStudy`. I'm wondering if we should update the `searchTraits` function in `src/data-sources/intermine/api/search-traits.ts` to handle this scenario, i.e. filter out such results. For example:
```typescript
    if (studyType == "GWAS") {
        const constraint = intermineNotNullConstraint("Trait.gwas");
        constraints.push(constraint);
    } else if (studyType == "QTLStudy") {
        const constraint = intermineNotNullConstraint("Trait.qtlStudy");
        constraints.push(constraint);
    } else {
        const constraint = intermineNotNullConstraint("Trait.gwas", "A");
        constraints.push(constraint);
        const constraint = intermineNotNullConstraint("Trait.qtlStudy", "B");
        constraints.push(constraint);
    }
    // ...
    const query = interminePathQuery(
        intermineTraitAttributes,
        intermineTraitSort,
        constraints,
        [],
        "A or B",
    );
```
Can @sammyjava or @adf-ncgr shed some light on this?

**Comment by sammyjava** [comment]
Traits currently only come from QTL studies or GWAS. Not that this wouldn't be useful in general. But what you say isn't true.
```
minimine-5103=> select primaryidentifier,qtlstudyid,gwasid from trait order by qtlstudyid,gwasid;
                        primaryidentifier                         | qtlstudyid |  gwasid  
------------------------------------------------------------------+------------+----------
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Plant_width              |   47000006 |         
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Yield                    |   47000006 |         
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Plant_height             |   47000006 |         
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Pods_per_plant           |   47000006 |         
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Days_to_maturity         |   47000006 |         
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Seeds_per_plant          |   47000006 |         
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Days_to_flowering        |   47000006 |         
 Cerinza_x_G24404.qtl.Blair_Iriarte_2006:Seed_weight              |   47000006 |         
 DOR364_x_BAT477.qtl.Blair_Galeano_2012:Yield_per_day             |   48000010 |         
 DOR364_x_BAT477.qtl.Blair_Galeano_2012:Days_to_flowering         |   48000010 |         
 DOR364_x_BAT477.qtl.Blair_Galeano_2012:Days_to_maturity          |   48000010 |         
 DOR364_x_BAT477.qtl.Blair_Galeano_2012:Seed_weight               |   48000010 |         
 DOR364_x_BAT477.qtl.Blair_Galeano_2012:Yield                     |   48000010 |         
 ZN016_x_Zhijiang282.qtl.Xu_Wu_2017:Pod_length                    |   94000014 |         
 mixed.gwas.Jain_Poromarto_2019:Soybean_cyst_nematode             |            | 45000008
 mixed.gwas.Myers_Wallace_2019:Immature_pod_color__blue_to_yellow |            | 46000012
 mixed.gwas.Myers_Wallace_2019:Immature_pod_color__lightness      |            | 46000012
 mixed.gwas.Myers_Wallace_2019:Immature_pod_color__green_to_red   |            | 46000012
 mixed.gwas.Myers_Wallace_2019:Flower_color                       |            | 46000012
 mixed.gwas.Myers_Wallace_2019:Pod_phenol                         |            | 46000012
 mixed.gwas.Raggi_Caproni_2019:Days_to_flowering                  |            | 92000007
 ZN016_x_Zhijiang282.gwas.Xu_Wu_2017:Pod_length                   |            | 93000015
```

**Comment by alancleary** [comment]
As discussed in https://github.com/legumeinfo/web-components/pull/267#discussion_r1401053679, this appears to be a legitimate bug having to do with how the GraphQL Server gets GWAS for a specific trait: https://github.com/legumeinfo/graphql-server/blob/main/src/data-sources/intermine/api/get-gwas-for-trait.ts#L19. @sammyjava any thoughts on how we might address this?

**Comment by sammyjava** [comment]
Yikes, it's actually missing data in GlycineMine, which explains why it didn't reproduce on MiniMine!
```
glycinemine-5103=> select * from trait where name='Branching' and gwasid is not null;
 description |    id     |   name    |         primaryidentifier         | qtlstudyid |  gwasid   | datasetid |             class             
-------------+-----------+-----------+-----------------------------------+------------+-----------+-----------+-------------------------------
             | 298000489 | Branching | mixed.gwas.Fang_Ma_2017:Branching |            | 298000034 | 298000001 | org.intermine.model.bio.Trait
(1 row)

glycinemine-5103=> select * from gwasresult where traitid =298000489;
 id | pvalue | primaryidentifier | markername | gwasid | traitid | datasetid | class 
----+--------+-------------------+------------+--------+---------+-----------+-------
(0 rows)
```
That second query is how `getGWASForTrait` finds the GWAS for a trait. (GWAS does not have a traits collection, it has a results collection and each result has a trait reference).

So I clearly have a bug in the mine GWAS loader that dropped that GWASResult-Trait association. The GWASResults for that GWAS  have trait ID from 298000037 to 298000444 and this trait is 298000489. Weird.

So I'll post that under `mine-issues` and close this out since it's not a `graphql-server` issue. I'll let you'all know when it's fixed and GlycineMine has been updated.

**Comment by alancleary** [comment]
Thanks for digging into this @sammyjava. Hopefully it's not too difficult to resolve in the mine code!

**Comment by sammyjava** [comment]
I figured it out, it's a data bug. The obo file in that collection includes Branching, but there is no Branching trait in the results file. So the trait got created from an errant obo file, which I've fixed. I'll yank that errant trait from the mine. I build those obo files myself, manually, and this is probably not the most unlikely data error to happen.

I thought the gwas collection validator checks that the trait names in the obo are present in the results file and vice versa, but apparently it failed, so that's the actual update out of all this.
